### PR TITLE
Use PostgreSQL's GIN index and `jsonb_path_ops` to form queries

### PR DESCRIPTION
**Pull request by rspeer**
This branch reworks many kinds of ConceptNet queries so that they use the JSONB `@>` operator. Queries that use this operator can be indexed efficiently using GIN indexing, and specifically the `jsonb_path_ops` index type.

Documentation of GIN is very sparse, but the most relevant place it's described is https://www.postgresql.org/docs/11/datatype-json.html .

This lets us construct SQL queries out of the user-supplied criteria much more naturally, and also fixes some terrible worst-case performance by allowing us to skip messy joins against the `node_prefixes` table that Postgres would sometimes formulate a really bad query plan for. (In fact, that table is removed, because we no longer have a need for it.)

We also skip gzipping the PostgreSQL CSV files. (We just need to rsync them to the server, and rsync already knows how to gzip.)

### CODE REVIEW

File: conceptnet5/db/prepare_data.py [hunk 1]
```diff
@@ -28,24 +28,54 @@ def sanitize(text):
...
+        'end': ['/c/en', '/c/en/bark'],
+        'rel': ['/r/CapableOf'],
+        ...
+    }
+    """
+    gin_edge = {}
+    gin_edge['uri'] = edge['uri']
+    gin_edge['start'] = uri_prefixes(edge['start'])
+    gin_edge['end'] = uri_prefixes(edge['end'])
+    gin_edge['rel'] = uri_prefixes(edge['rel'])
```
Comment (jlowryduda) [inline conceptnet5/db/prepare_data.py]: I think `uri_prefixes` will always just return the relation itself, for example `/r/CapableOf`, so it should be enough to say `gin_edge['rel'] = edge['rel']`.

File: conceptnet5/db/prepare_data.py [hunk 2]
```diff
@@ -71,45 +101,48 @@ def assertions_to_sql_csv(msgpack_filename, output_dir):
...
-        for source_idx in sorted(set(source_indices)):
-            write_row(edge_source_file, [assertion_idx, source_idx])
+        write_row(
+            edge_gin_file,
+            [assertion_idx, weight,
+             json.dumps(gin_indexable_edge(assertion), ensure_ascii=False)]
+        )
 
+        features = []
+        start_p_indices = [
```
Comment (jlowryduda) [inline conceptnet5/db/prepare_data.py]: Could we get more documentation here?

File: conceptnet5/db/query.py [hunk 3]
```diff
@@ -1,156 +1,147 @@
...
+        ORDER BY random() LIMIT %(limit)s
+    """
 
+# A query that's optimized for producing the edges, grouped by feature, that
+# you get when you look up a concept in the Web interface.
 NODE_TO_FEATURE_QUERY = """
 WITH node_ids AS (
-    SELECT p.node_id FROM nodes n, node_prefixes p
-    WHERE p.prefix_id=n.id AND n.uri=%(node)s
-    LIMIT 10
```
Comment (jlowryduda) [inline conceptnet5/db/query.py]: Why is there no limit here anymore?

File: conceptnet5/db/query.py [hunk 4]
```diff
@@ -1,156 +1,147 @@
...
-MAX_GROUP_SIZE = 20
+
+# Queries that match arbitrary criteria using a GIN index. The @> operator
+# tests whether one JSONB structure includes all the values in another.
+GIN_QUERY_1WAY = """
+WITH matched_edges AS (
+    SELECT edge_id FROM edges_gin
+    WHERE data @> %(query)s
+    LIMIT 10000
+)
```
Comment (jlowryduda) [inline conceptnet5/db/query.py]: Whoa, that's very nice and elegant.

File: conceptnet5/db/prepare_data.py (hunk 1 above)
Comment (rspeer) [reply conceptnet5/db/prepare_data.py]: It still needs to be a list, and we might as well be consistent. This additionally allows matching all the DBPedia relations that are "on probation" in their own namespace, like `/r/dbpedia/genre`, by querying for `/r/dbpedia`.

File: conceptnet5/db/query.py (hunk 3 above)
Comment (rspeer) [reply conceptnet5/db/query.py]: It was kind of a desperate optimization. If the 'node' matched too many things, we'd just return an arbitrary 10 of them.

Now that we're looking up the prefix and the features are stored separately for each prefix, we don't need to worry about this.

**Comment by rspeer** [comment]
I worked on cleaning up `prepare_data`, and I wondered why we were throwing out duplicate assertions, given that assertions were supposed to be "edges combined together so that they're unique on their `(rel, start, end)` triple" and therefore there shouldn't be any duplicates.

I found the reason: because we reduce `start` and `end` to their prefix of 4 URI pieces, throwing away word sense information, when we made assertions, and that was making distinct edges become the same assertion. And we were just throwing one of them out.

I think this was another one of those desperate shortcuts, so I'm seeing what happens if I undo it and actually preserve the word sense information.

**Pull request merged**